diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index 729b4c3fa..036fc9b91 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -4030,7 +4030,6 @@ bool ssl_add_clienthello_tlsext(SSL_HANDSHAKE *hs, CBB *out, CBB *out_encoded,
   }
 
   bool last_was_empty = false;
-  bool hasPreSharedKey = false;
   const size_t numExtensions = hs->extension_permutation.empty()
       ? kNumExtensions
       : hs->extension_permutation.size();
@@ -4038,11 +4037,6 @@ bool ssl_add_clienthello_tlsext(SSL_HANDSHAKE *hs, CBB *out, CBB *out_encoded,
     size_t i = hs->extension_permutation.empty()
                    ? unpermuted
                    : hs->extension_permutation[unpermuted];
-    if (i == 126)
-    {
-        hasPreSharedKey = true;
-        continue;
-    }
     if (i >= kNumExtensions) {
         continue;
     }
@@ -4118,17 +4112,11 @@ bool ssl_add_clienthello_tlsext(SSL_HANDSHAKE *hs, CBB *out, CBB *out_encoded,
       return false;
     }
   }
-
-   if (hasPreSharedKey)
-   {
-       // The PSK extension must be last, including after the padding.
-       size_t psk_len_actual;
-       if (!ext_pre_shared_key_add_clienthello(hs, out, &extensions, &psk_len_actual,
-           type)) {
-           return false;
-       }
-}
-   assert(psk_len_actual == psk_len);
+  size_t psk_len_actual;
+   if (!ext_pre_shared_key_add_clienthello(hs, out, &extensions, &psk_len_actual,
+       type)) {
+       return false;
+   }
    return true;
 }
 
diff --git a/ssl/ssl_cipher.cc b/ssl/ssl_cipher.cc
index a10329bc4..1c4fb2cf9 100644
--- a/ssl/ssl_cipher.cc
+++ b/ssl/ssl_cipher.cc
@@ -338,16 +338,6 @@ static constexpr SSL_CIPHER kCiphers[] = {
               // Client Hello.
               // HMAC based TLS v1.2 ciphersuites from RFC5289
               // Cipher C023
-    {
-    TLS1_TXT_ECDHE_ECDSA_WITH_AES_128_SHA256,
-    "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
-    TLS1_CK_ECDHE_ECDSA_WITH_AES_128_SHA256,
-    SSL_kECDHE,
-    SSL_aECDSA,
-    SSL_AES128,
-    SSL_SHA256,
-    SSL_HANDSHAKE_MAC_SHA256,
-    },
 
     // Cipher C024
 
@@ -1366,18 +1356,18 @@ static int ssl_cipher_id_cmp_void(const void *in_a, const void *in_b) {
                            reinterpret_cast<const SSL_CIPHER *>(in_b));
 }
 
-//template <size_t N>
-//static constexpr bool ssl_ciphers_sorted(const SSL_CIPHER (&ciphers)[N]) {
-//  for (size_t i = 1; i < N; i++) {
-//    if (ssl_cipher_id_cmp(&ciphers[i - 1], &ciphers[i]) >= 0) {
-//      return false;
-//    }
-//  }
-//  return true;
-//}
-//
-//static_assert(ssl_ciphers_sorted(kCiphers),
-//              "Ciphers are not sorted, bsearch won't work");
+template <size_t N>
+static constexpr bool ssl_ciphers_sorted(const SSL_CIPHER (&ciphers)[N]) {
+  for (size_t i = 1; i < N; i++) {
+    if (ssl_cipher_id_cmp(&ciphers[i - 1], &ciphers[i]) >= 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static_assert(ssl_ciphers_sorted(kCiphers),
+              "Ciphers are not sorted, bsearch won't work");
 
 const SSL_CIPHER *SSL_get_cipher_by_value(uint16_t value) {
   SSL_CIPHER c;
