diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
index 839cf2f73..5b0164882 100644
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -5315,6 +5315,9 @@ OPENSSL_EXPORT int SSL_CTX_set1_curves_list(SSL_CTX *ctx, const char *curves);
 // SSL_set1_curves_list calls |SSL_set1_groups_list|.
 OPENSSL_EXPORT int SSL_set1_curves_list(SSL *ssl, const char *curves);
 
+OPENSSL_EXPORT int SSL_CTX_set_ciphersuites(SSL_CTX* ctx, char* str);
+
+OPENSSL_EXPORT int SSL_CTX_set_extension_permutation(SSL_CTX* ctx, char* str);
 // TLSEXT_nid_unknown is a constant used in OpenSSL for
 // |SSL_get_negotiated_group| to return an unrecognized group. BoringSSL never
 // returns this value, but we define this constant for compatibility.
diff --git a/include/openssl/tls1.h b/include/openssl/tls1.h
index c1207a3b7..602cc83d1 100644
--- a/include/openssl/tls1.h
+++ b/include/openssl/tls1.h
@@ -181,6 +181,7 @@ extern "C" {
 #define TLS1_AD_NO_APPLICATION_PROTOCOL 120
 #define TLS1_AD_ECH_REQUIRED 121  // draft-ietf-tls-esni-13
 
+#define TLSEXT_TYPE_record_size_limit 28
 // ExtensionType values from RFC 6066
 #define TLSEXT_TYPE_server_name 0
 #define TLSEXT_TYPE_status_request 5
diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index b13400097..07041c87d 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -420,8 +420,8 @@ static const uint16_t kSignSignatureAlgorithms[] = {
     SSL_SIGN_RSA_PKCS1_SHA512,
 
     // If the peer supports nothing else, sign with SHA-1.
-    SSL_SIGN_ECDSA_SHA1,
-    SSL_SIGN_RSA_PKCS1_SHA1,
+    //SSL_SIGN_ECDSA_SHA1,
+    //SSL_SIGN_RSA_PKCS1_SHA1,
 };
 
 static Span<const uint16_t> tls12_get_verify_sigalgs(const SSL_HANDSHAKE *hs) {
@@ -1148,20 +1148,20 @@ static bool ext_ocsp_add_serverhello(SSL_HANDSHAKE *hs, CBB *out) {
 static bool ext_npn_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
                                     CBB *out_compressible,
                                     ssl_client_hello_type_t type) {
-  const SSL *const ssl = hs->ssl;
-  if (ssl->ctx->next_proto_select_cb == NULL ||
-      // Do not allow NPN to change on renegotiation.
-      ssl->s3->initial_handshake_complete ||
-      // NPN is not defined in DTLS or TLS 1.3.
-      SSL_is_dtls(ssl) || hs->min_version >= TLS1_3_VERSION ||
-      type == ssl_client_hello_inner) {
-    return true;
-  }
-
-  if (!CBB_add_u16(out, TLSEXT_TYPE_next_proto_neg) ||
-      !CBB_add_u16(out, 0 /* length */)) {
-    return false;
-  }
+  //const SSL *const ssl = hs->ssl;
+  //if (ssl->ctx->next_proto_select_cb == NULL ||
+  //    // Do not allow NPN to change on renegotiation.
+  //    ssl->s3->initial_handshake_complete ||
+  //    // NPN is not defined in DTLS or TLS 1.3.
+  //    SSL_is_dtls(ssl) || hs->min_version >= TLS1_3_VERSION ||
+  //    type == ssl_client_hello_inner) {
+  //  return true;
+  //}
+
+  //if (!CBB_add_u16(out, TLSEXT_TYPE_next_proto_neg) ||
+  //    !CBB_add_u16(out, 0 /* length */)) {
+  //  return false;
+  //}
 
   return true;
 }
@@ -2757,6 +2757,24 @@ static bool ext_quic_transport_params_add_serverhello_legacy(SSL_HANDSHAKE *hs,
 static bool ext_delegated_credential_add_clienthello(
     const SSL_HANDSHAKE *hs, CBB *out, CBB *out_compressible,
     ssl_client_hello_type_t type) {
+    CBB contents, data;
+    static const uint16_t signature_hash_algorithms[] = {
+    SSL_SIGN_ECDSA_SECP256R1_SHA256, SSL_SIGN_ECDSA_SECP384R1_SHA384,
+    SSL_SIGN_ECDSA_SECP521R1_SHA512, SSL_SIGN_ECDSA_SHA1 };
+    if (!CBB_add_u16(out, TLSEXT_TYPE_delegated_credential) ||
+            !CBB_add_u16_length_prefixed(out, &contents) ||
+            !CBB_add_u16_length_prefixed(&contents, &data)) {
+        return false;
+    }
+    
+    for (const uint16_t alg : signature_hash_algorithms) {
+        if (!CBB_add_u16(&data, alg)) {
+            return false;
+        }
+    }
+    if (!CBB_flush(out)) {
+        return false;
+    }
   return true;
 }
 
@@ -2950,12 +2968,12 @@ static bool ext_alps_add_clienthello_impl(const SSL_HANDSHAKE *hs, CBB *out,
   return CBB_flush(out_compressible);
 }
 
-static bool ext_alps_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
-                                     CBB *out_compressible,
-                                     ssl_client_hello_type_t type) {
-  return ext_alps_add_clienthello_impl(hs, out, out_compressible, type,
-                                       /*use_new_codepoint=*/true);
-}
+//static bool ext_alps_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
+//                                     CBB *out_compressible,
+//                                     ssl_client_hello_type_t type) {
+//  return ext_alps_add_clienthello_impl(hs, out, out_compressible, type,
+//                                       /*use_new_codepoint=*/true);
+//}
 
 static bool ext_alps_add_clienthello_old(const SSL_HANDSHAKE *hs, CBB *out,
                                          CBB *out_compressible,
@@ -2996,12 +3014,12 @@ static bool ext_alps_parse_serverhello_impl(SSL_HANDSHAKE *hs,
   return true;
 }
 
-static bool ext_alps_parse_serverhello(SSL_HANDSHAKE *hs,
-                                       uint8_t *out_alert,
-                                       CBS *contents) {
-  return ext_alps_parse_serverhello_impl(hs, out_alert, contents,
-                                         /*use_new_codepoint=*/true);
-}
+//static bool ext_alps_parse_serverhello(SSL_HANDSHAKE *hs,
+//                                       uint8_t *out_alert,
+//                                       CBS *contents) {
+//  return ext_alps_parse_serverhello_impl(hs, out_alert, contents,
+//                                         /*use_new_codepoint=*/true);
+//}
 
 static bool ext_alps_parse_serverhello_old(SSL_HANDSHAKE *hs,
                                            uint8_t *out_alert,
@@ -3044,9 +3062,9 @@ static bool ext_alps_add_serverhello_impl(SSL_HANDSHAKE *hs, CBB *out,
   return true;
 }
 
-static bool ext_alps_add_serverhello(SSL_HANDSHAKE *hs, CBB *out) {
-  return ext_alps_add_serverhello_impl(hs, out, /*use_new_codepoint=*/true);
-}
+//static bool ext_alps_add_serverhello(SSL_HANDSHAKE *hs, CBB *out) {
+//  return ext_alps_add_serverhello_impl(hs, out, /*use_new_codepoint=*/true);
+//}
 
 static bool ext_alps_add_serverhello_old(SSL_HANDSHAKE *hs, CBB *out) {
   return ext_alps_add_serverhello_impl(hs, out, /*use_new_codepoint=*/false);
@@ -3108,6 +3126,35 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
   return true;
 }
 
+
+static bool record_size_limit_add_clienthello(const SSL_HANDSHAKE* hs, CBB* out,
+    CBB* out_compressible,
+    ssl_client_hello_type_t type) {
+    CBB data;
+    const uint16_t data_ = 0x4001;
+    if (!CBB_add_u16(out, TLSEXT_TYPE_record_size_limit) ||
+        !CBB_add_u16_length_prefixed(out, &data) || !CBB_add_u16(&data, data_) ||
+        !CBB_flush(out)) {
+        return false;
+    }
+    return true;
+}
+
+static bool record_size_limit_parse_serverhello(SSL_HANDSHAKE* hs,
+    uint8_t* out_alert,
+    CBS* contents) {
+    return true;
+}
+
+static bool record_size_limit_parse_clienthello(SSL_HANDSHAKE* hs,
+    uint8_t* out_alert,
+    CBS* contents) {
+    return true;
+}
+
+static bool record_size_limit_add_serverhello(SSL_HANDSHAKE* hs, CBB* out) {
+    return true;
+}
 // kExtensions contains all the supported extensions.
 static const struct tls_extension kExtensions[] = {
   {
@@ -3273,14 +3320,14 @@ static const struct tls_extension kExtensions[] = {
     ext_delegated_credential_parse_clienthello,
     dont_add_serverhello,
   },
-  {
-    TLSEXT_TYPE_application_settings,
-    ext_alps_add_clienthello,
-    ext_alps_parse_serverhello,
-    // ALPS is negotiated late in |ssl_negotiate_alpn|.
-    ignore_parse_clienthello,
-    ext_alps_add_serverhello,
-  },
+  //{
+  //  TLSEXT_TYPE_application_settings,
+  //  ext_alps_add_clienthello,
+  //  ext_alps_parse_serverhello,
+  //  // ALPS is negotiated late in |ssl_negotiate_alpn|.
+  //  ignore_parse_clienthello,
+  //  ext_alps_add_serverhello,
+  //},
   {
     TLSEXT_TYPE_application_settings_old,
     ext_alps_add_clienthello_old,
@@ -3289,6 +3336,13 @@ static const struct tls_extension kExtensions[] = {
     ignore_parse_clienthello,
     ext_alps_add_serverhello_old,
   },
+  {
+    TLSEXT_TYPE_record_size_limit,
+    record_size_limit_add_clienthello,
+    record_size_limit_parse_serverhello,
+    record_size_limit_parse_clienthello,
+    record_size_limit_add_serverhello,
+  },
 };
 
 #define kNumExtensions (sizeof(kExtensions) / sizeof(struct tls_extension))
@@ -3302,6 +3356,12 @@ static_assert(kNumExtensions <=
 
 bool ssl_setup_extension_permutation(SSL_HANDSHAKE *hs) {
   if (!hs->config->permute_extensions) {
+      if (!hs->ssl->ctx->extension_permutation.empty()) {
+          Array<uint8_t> permutation;
+          permutation.Init(hs->ssl->ctx->extension_permutation.size());
+          permutation.CopyFrom(hs->ssl->ctx->extension_permutation);
+          hs->extension_permutation = std::move(permutation);
+      }
     return true;
   }
 
@@ -3378,11 +3438,16 @@ static bool ssl_add_clienthello_tlsext_inner(SSL_HANDSHAKE *hs, CBB *out,
       return false;
     }
   }
-
-  for (size_t unpermuted = 0; unpermuted < kNumExtensions; unpermuted++) {
+  const size_t numExtensions = hs->extension_permutation.empty()
+      ? kNumExtensions
+      : hs->extension_permutation.size();
+  for (size_t unpermuted = 0; unpermuted < numExtensions; unpermuted++) {
     size_t i = hs->extension_permutation.empty()
                    ? unpermuted
                    : hs->extension_permutation[unpermuted];
+    if (i >= kNumExtensions) {
+        continue;
+    }
     const size_t len_before = CBB_len(&extensions);
     const size_t len_compressed_before = CBB_len(compressed.get());
     if (!kExtensions[i].add_clienthello(hs, &extensions, compressed.get(),
@@ -3440,19 +3505,27 @@ static bool ssl_add_clienthello_tlsext_inner(SSL_HANDSHAKE *hs, CBB *out,
     }
   }
 
+  bool hasPreSharedKey = false;
+  for (const size_t id : hs->extension_permutation) {
+      if (id >= kNumExtensions) {
+          hasPreSharedKey = true;
+          break;
+      }
+  }
   // The PSK extension must be last. It is never compressed. Note, if there is a
   // binder, the caller will need to update both ClientHelloInner and
   // EncodedClientHelloInner after computing it.
-  const size_t len_before = CBB_len(&extensions);
-  if (!ext_pre_shared_key_add_clienthello(hs, &extensions, out_needs_psk_binder,
-                                          ssl_client_hello_inner) ||
-      !CBB_add_bytes(&extensions_encoded, CBB_data(&extensions) + len_before,
-                     CBB_len(&extensions) - len_before) ||
-      !CBB_flush(out) ||  //
-      !CBB_flush(out_encoded)) {
-    return false;
+  if (hasPreSharedKey) {
+      const size_t len_before = CBB_len(&extensions);
+      if (!ext_pre_shared_key_add_clienthello(hs, &extensions, out_needs_psk_binder,
+          ssl_client_hello_inner) ||
+          !CBB_add_bytes(&extensions_encoded, CBB_data(&extensions) + len_before,
+              CBB_len(&extensions) - len_before) ||
+          !CBB_flush(out) ||  //
+          !CBB_flush(out_encoded)) {
+          return false;
+      }
   }
-
   return true;
 }
 
@@ -3488,10 +3561,16 @@ bool ssl_add_clienthello_tlsext(SSL_HANDSHAKE *hs, CBB *out, CBB *out_encoded,
   }
 
   bool last_was_empty = false;
-  for (size_t unpermuted = 0; unpermuted < kNumExtensions; unpermuted++) {
+  const size_t numExtensions = hs->extension_permutation.empty()
+      ? kNumExtensions
+      : hs->extension_permutation.size();
+  for (size_t unpermuted = 0; unpermuted < numExtensions; unpermuted++) {
     size_t i = hs->extension_permutation.empty()
                    ? unpermuted
                    : hs->extension_permutation[unpermuted];
+    if (i >= kNumExtensions) {
+        continue;
+    }
     const size_t len_before = CBB_len(&extensions);
     if (!kExtensions[i].add_clienthello(hs, &extensions, &extensions, type)) {
       OPENSSL_PUT_ERROR(SSL, SSL_R_ERROR_ADDING_EXTENSION);
@@ -3564,15 +3643,26 @@ bool ssl_add_clienthello_tlsext(SSL_HANDSHAKE *hs, CBB *out, CBB *out_encoded,
     }
   }
 
-  // The PSK extension must be last, including after the padding.
-  const size_t len_before = CBB_len(&extensions);
-  if (!ext_pre_shared_key_add_clienthello(hs, &extensions, out_needs_psk_binder,
-                                          type)) {
-    OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
-    return false;
+  bool hasPreSharedKey = false;
+  for (const size_t id : hs->extension_permutation) {
+      if (id >= kNumExtensions) {
+          hasPreSharedKey = true;
+          break;
+      }
+  }
+  if (hasPreSharedKey)
+  {
+      // The PSK extension must be last, including after the padding.
+      const size_t len_before = CBB_len(&extensions);
+      if (!ext_pre_shared_key_add_clienthello(hs, &extensions, out_needs_psk_binder,
+          type)) {
+          OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+          return false;
+      }
+      assert(psk_extension_len == CBB_len(&extensions) - len_before);
+      (void)len_before;  // |assert| is omitted in release builds.
   }
-  assert(psk_extension_len == CBB_len(&extensions) - len_before);
-  (void)len_before;  // |assert| is omitted in release builds.
+  
 
   // Discard empty extensions blocks.
   if (CBB_len(&extensions) == 0) {
diff --git a/ssl/handshake_client.cc b/ssl/handshake_client.cc
index 971ebd0b1..1fc406f98 100644
--- a/ssl/handshake_client.cc
+++ b/ssl/handshake_client.cc
@@ -243,24 +243,48 @@ static bool ssl_write_client_cipher_list(const SSL_HANDSHAKE *hs, CBB *out,
   // Add TLS 1.3 ciphers. Order ChaCha20-Poly1305 relative to AES-GCM based on
   // hardware support.
   if (hs->max_version >= TLS1_3_VERSION) {
-    const bool has_aes_hw = ssl->config->aes_hw_override
-                                ? ssl->config->aes_hw_override_value
-                                : EVP_has_aes_hardware();
-
-    if ((!has_aes_hw &&  //
-         !ssl_add_tls13_cipher(&child,
-                               TLS1_3_CK_CHACHA20_POLY1305_SHA256 & 0xffff,
-                               ssl->config->tls13_cipher_policy)) ||
-        !ssl_add_tls13_cipher(&child, TLS1_3_CK_AES_128_GCM_SHA256 & 0xffff,
-                              ssl->config->tls13_cipher_policy) ||
-        !ssl_add_tls13_cipher(&child, TLS1_3_CK_AES_256_GCM_SHA384 & 0xffff,
-                              ssl->config->tls13_cipher_policy) ||
-        (has_aes_hw &&  //
-         !ssl_add_tls13_cipher(&child,
-                               TLS1_3_CK_CHACHA20_POLY1305_SHA256 & 0xffff,
-                               ssl->config->tls13_cipher_policy))) {
-      return false;
+    if (!ssl->ctx->cipher13_flag_list.empty()) {
+        for (uint8_t flag : ssl->ctx->cipher13_flag_list) {
+            int id = 0;
+            switch (flag) {
+            case 1:
+                id = TLS1_3_CK_AES_128_GCM_SHA256;
+                break;
+            case 2:
+                id = TLS1_3_CK_AES_256_GCM_SHA384;
+                break;
+            case 3:
+                id = TLS1_3_CK_CHACHA20_POLY1305_SHA256;
+                break;
+            default:
+                return false;
+            }
+            if (!CBB_add_u16(&child, id & 0xffff)) {
+                return false;
+            }
+        }
+    }
+    else {
+        const bool has_aes_hw = ssl->config->aes_hw_override
+            ? ssl->config->aes_hw_override_value
+            : EVP_has_aes_hardware();
+
+        if ((!has_aes_hw &&  //
+            !ssl_add_tls13_cipher(&child,
+                TLS1_3_CK_CHACHA20_POLY1305_SHA256 & 0xffff,
+                ssl->config->tls13_cipher_policy)) ||
+            !ssl_add_tls13_cipher(&child, TLS1_3_CK_AES_128_GCM_SHA256 & 0xffff,
+                ssl->config->tls13_cipher_policy) ||
+            !ssl_add_tls13_cipher(&child, TLS1_3_CK_AES_256_GCM_SHA384 & 0xffff,
+                ssl->config->tls13_cipher_policy) ||
+            (has_aes_hw &&  //
+                !ssl_add_tls13_cipher(&child,
+                    TLS1_3_CK_CHACHA20_POLY1305_SHA256 & 0xffff,
+                    ssl->config->tls13_cipher_policy))) {
+            return false;
+        }
     }
+    
   }
 
   if (hs->min_version < TLS1_3_VERSION && type != ssl_client_hello_inner) {
diff --git a/ssl/internal.h b/ssl/internal.h
index d9972dc83..f929dfb9f 100644
--- a/ssl/internal.h
+++ b/ssl/internal.h
@@ -563,8 +563,9 @@ BSSL_NAMESPACE_BEGIN
 // Bits for |algorithm_mac| (symmetric authentication).
 #define SSL_SHA1 0x00000001u
 #define SSL_SHA256 0x00000002u
+#define SSL_SHA384 0x00000004u
 // SSL_AEAD is set for all AEADs.
-#define SSL_AEAD 0x00000004u
+#define SSL_AEAD 0x00000008u
 
 // Bits for |algorithm_prf| (handshake digest).
 #define SSL_HANDSHAKE_MAC_DEFAULT 0x1
@@ -3754,6 +3755,8 @@ struct ssl_ctx_st {
   // |aes_hw_override| is true.
   bool aes_hw_override_value : 1;
 
+  bssl::Array<uint8_t> cipher13_flag_list;
+  bssl::Array<uint8_t> extension_permutation;
  private:
   ~ssl_ctx_st();
   friend OPENSSL_EXPORT void SSL_CTX_free(SSL_CTX *);
diff --git a/ssl/ssl_cipher.cc b/ssl/ssl_cipher.cc
index fd8cef95d..c5f86041b 100644
--- a/ssl/ssl_cipher.cc
+++ b/ssl/ssl_cipher.cc
@@ -197,6 +197,64 @@ static constexpr SSL_CIPHER kCiphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT,
     },
 
+    // Cipher 39
+    {
+    TLS1_TXT_DHE_RSA_WITH_AES_256_SHA,
+    "TLS_DHE_RSA_WITH_AES_256_SHA",
+    TLS1_CK_DHE_RSA_WITH_AES_256_SHA,
+    SSL_kRSA,
+    SSL_aRSA,
+    SSL_AES256,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+    },
+    
+        // Cipher 3C
+    {
+    TLS1_TXT_RSA_WITH_AES_128_SHA256,
+    "TLS_RSA_WITH_AES_128_SHA256",
+    TLS1_CK_RSA_WITH_AES_128_SHA256,
+    SSL_kRSA,
+    SSL_aRSA,
+    SSL_AES256,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+    },
+        // Cipher 3D
+    {
+    TLS1_TXT_RSA_WITH_AES_256_SHA256,
+    "TLS_RSA_WITH_AES_256_SHA256",
+    TLS1_CK_RSA_WITH_AES_256_SHA256,
+    SSL_kRSA,
+    SSL_aRSA,
+    SSL_AES256,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+    },
+    
+        // Cipher 67
+    {
+    TLS1_TXT_DHE_RSA_WITH_AES_128_SHA256,
+    "TLS_DHE_RSA_WITH_AES_128_SHA256",
+    TLS1_CK_DHE_RSA_WITH_AES_128_SHA256,
+    SSL_kRSA,
+    SSL_aRSA,
+    SSL_AES128,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+    },
+    
+        // Cipher 6B
+    {
+    TLS1_TXT_DHE_RSA_WITH_AES_256_SHA256,
+    "TLS_DHE_RSA_WITH_AES_256_SHA256",
+    TLS1_CK_DHE_RSA_WITH_AES_256_SHA256,
+    SSL_kRSA,
+    SSL_aRSA,
+    SSL_AES256,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+     },
     // PSK cipher suites.
 
     // Cipher 8C
@@ -248,7 +306,29 @@ static constexpr SSL_CIPHER kCiphers[] = {
      SSL_AEAD,
      SSL_HANDSHAKE_MAC_SHA384,
     },
-
+        // Cipher 9E
+    {
+    TLS1_TXT_DHE_RSA_WITH_AES_128_GCM_SHA256,
+    "TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
+    TLS1_CK_DHE_RSA_WITH_AES_128_GCM_SHA256,
+    SSL_kECDHE,
+    SSL_aGENERIC,
+    SSL_AES128GCM,
+    SSL_AEAD,
+    SSL_HANDSHAKE_MAC_SHA256,
+    },
+        
+    // Cipher 9F
+    {
+    TLS1_TXT_DHE_RSA_WITH_AES_256_GCM_SHA384,
+    "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
+    TLS1_CK_DHE_RSA_WITH_AES_256_GCM_SHA384,
+    SSL_kECDHE,
+    SSL_aGENERIC,
+    SSL_AES256GCM,
+    SSL_AEAD,
+    SSL_HANDSHAKE_MAC_SHA384,
+    },
     // TLS 1.3 suites.
 
     // Cipher 1301
@@ -287,6 +367,17 @@ static constexpr SSL_CIPHER kCiphers[] = {
       SSL_HANDSHAKE_MAC_SHA256,
     },
 
+    // Cipher C008
+    {
+    "ECDHE-ECDSA-DES-CBC3-SHA",
+    "TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA",
+    0x0300C008,
+    SSL_kECDHE,
+    SSL_aECDSA,
+    SSL_3DES,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+    },
     // Cipher C009
     {
      TLS1_TXT_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
@@ -311,6 +402,17 @@ static constexpr SSL_CIPHER kCiphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT,
     },
 
+    // Cipher C012
+    {
+    "ECDHE-RSA-DES-CBC3-SHA",
+    "TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA",
+    0x0300C012,
+    SSL_kECDHE,
+    SSL_aRSA,
+    SSL_3DES,
+    SSL_SHA1,
+    SSL_HANDSHAKE_MAC_DEFAULT,
+     },
     // Cipher C013
     {
      TLS1_TXT_ECDHE_RSA_WITH_AES_128_CBC_SHA,
@@ -335,6 +437,37 @@ static constexpr SSL_CIPHER kCiphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT,
     },
 
+       // curl-impersonate: Ciphers C023, C024, C027, C028 were removed in
+      // https://boringssl-review.googlesource.com/c/boringssl/+/27944/
+              // but restored here to impersonate browsers with older ciphers. They
+              // are
+              // not expected to actually work; but just to be included in the TLS
+              // Client Hello.
+              // HMAC based TLS v1.2 ciphersuites from RFC5289
+              // Cipher C023
+    {
+    TLS1_TXT_ECDHE_ECDSA_WITH_AES_128_SHA256,
+    "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
+    TLS1_CK_ECDHE_ECDSA_WITH_AES_128_SHA256,
+    SSL_kECDHE,
+    SSL_aECDSA,
+    SSL_AES128,
+    SSL_SHA256,
+    SSL_HANDSHAKE_MAC_SHA256,
+    },
+          
+    // Cipher C024
+              
+    {
+    TLS1_TXT_ECDHE_ECDSA_WITH_AES_256_SHA384,
+    "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384",
+    TLS1_CK_ECDHE_ECDSA_WITH_AES_256_SHA384,
+    SSL_kECDHE,
+    SSL_aECDSA,
+    SSL_AES256,
+    SSL_SHA384,
+    SSL_HANDSHAKE_MAC_SHA384,
+    },
     // Cipher C027
     {
      TLS1_TXT_ECDHE_RSA_WITH_AES_128_CBC_SHA256,
@@ -346,6 +479,17 @@ static constexpr SSL_CIPHER kCiphers[] = {
      SSL_SHA256,
      SSL_HANDSHAKE_MAC_SHA256,
     },
+    // Cipher C028
+    {
+    TLS1_TXT_ECDHE_RSA_WITH_AES_256_SHA384,
+    "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384",
+    TLS1_CK_ECDHE_RSA_WITH_AES_256_SHA384,
+    SSL_kECDHE,
+    SSL_aRSA,
+    SSL_AES256,
+    SSL_SHA384,
+    SSL_HANDSHAKE_MAC_SHA384,
+    },
 
     // GCM based TLS v1.2 ciphersuites from RFC 5289
 
@@ -467,15 +611,15 @@ Span<const SSL_CIPHER> AllCiphers() {
   return MakeConstSpan(kCiphers, OPENSSL_ARRAY_SIZE(kCiphers));
 }
 
-static constexpr size_t NumTLS13Ciphers() {
-  size_t num = 0;
-  for (const auto &cipher : kCiphers) {
-    if (cipher.algorithm_mkey == SSL_kGENERIC) {
-      num++;
-    }
-  }
-  return num;
-}
+//static constexpr size_t NumTLS13Ciphers() {
+//  size_t num = 0;
+//  for (const auto &cipher : kCiphers) {
+//    if (cipher.algorithm_mkey == SSL_kGENERIC) {
+//      num++;
+//    }
+//  }
+//  return num;
+//}
 
 #define CIPHER_ADD 1
 #define CIPHER_KILL 2
@@ -567,8 +711,11 @@ static const CIPHER_ALIAS kCipherAliases[] = {
 
     // Temporary no-op aliases corresponding to removed SHA-2 legacy CBC
     // ciphers. These should be removed after 2018-05-14.
-    {"SHA256", 0, 0, 0, 0, 0},
-    {"SHA384", 0, 0, 0, 0, 0},
+    /*{"SHA256", 0, 0, 0, 0, 0},
+    {"SHA384", 0, 0, 0, 0, 0},*/
+
+    {"SHA256", ~0u, ~0u, ~0u, SSL_SHA256, 0},
+    {"SHA384", ~0u, ~0u, ~0u, SSL_SHA384, 0},
 };
 
 static const size_t kCipherAliasesLen = OPENSSL_ARRAY_SIZE(kCipherAliases);
@@ -1210,9 +1357,9 @@ bool ssl_create_cipher_list(UniquePtr<SSLCipherPreferenceList> *out_cipher_list,
     assert(co_list[num - 1].cipher != nullptr);
   }
   assert(num == OPENSSL_ARRAY_SIZE(co_list));
-  static_assert(OPENSSL_ARRAY_SIZE(co_list) + NumTLS13Ciphers() ==
+  /*static_assert(OPENSSL_ARRAY_SIZE(co_list) + NumTLS13Ciphers() ==
                     OPENSSL_ARRAY_SIZE(kCiphers),
-                "Not all ciphers are included in the cipher order");
+                "Not all ciphers are included in the cipher order");*/
 
   // If the rule_string begins with DEFAULT, apply the default rule before
   // using the (possibly available) additional rules.
diff --git a/ssl/ssl_key_share.cc b/ssl/ssl_key_share.cc
index 80317d898..1d829f520 100644
--- a/ssl/ssl_key_share.cc
+++ b/ssl/ssl_key_share.cc
@@ -289,6 +289,10 @@ constexpr NamedGroup kNamedGroups[] = {
     {NID_X25519, SSL_GROUP_X25519, "X25519", "x25519"},
     {NID_X25519Kyber768Draft00, SSL_GROUP_X25519_KYBER768_DRAFT00,
      "X25519Kyber768Draft00", ""},
+
+    //unspport group but add them
+    {NID_secp224r1, 256, "dhe2048", "ffdhe2048"},
+    {NID_secp224r1, 257, "dhe3072", "ffdhe3072"},
 };
 
 }  // namespace
diff --git a/ssl/ssl_lib.cc b/ssl/ssl_lib.cc
index 58b68e675..a9bc96525 100644
--- a/ssl/ssl_lib.cc
+++ b/ssl/ssl_lib.cc
@@ -686,6 +686,13 @@ SSL *SSL_new(SSL_CTX *ctx) {
   ssl->config->ocsp_stapling_enabled = ctx->ocsp_stapling_enabled;
   ssl->config->handoff = ctx->handoff;
   ssl->quic_method = ctx->quic_method;
+  SSL_enable_ocsp_stapling(ssl.get());
+  SSL_add_application_settings(ssl.get(),
+      reinterpret_cast<const uint8_t*>("h2"),
+      2, NULL, 0);
+  SSL_CTX_set_grease_enabled(ssl->ctx.get(), 1);
+  SSL_enable_signed_cert_timestamps(ssl.get());
+  SSL_set_enable_ech_grease(ssl.get(), 1);
 
   if (!ssl->method->ssl_new(ssl.get()) ||
       !ssl->ctx->x509_method->ssl_new(ssl->s3->hs.get())) {
@@ -3353,7 +3360,6 @@ static int Configure(SSL_CTX *ctx) {
 
 static int Configure(SSL *ssl) {
   ssl->config->tls13_cipher_policy = ssl_compliance_policy_wpa3_192_202304;
-
   return SSL_set_min_proto_version(ssl, TLS1_2_VERSION) &&
          SSL_set_max_proto_version(ssl, TLS1_3_VERSION) &&
          SSL_set_strict_cipher_list(ssl, kTLS12Ciphers) &&
@@ -3388,3 +3394,33 @@ int SSL_set_compliance_policy(SSL *ssl, enum ssl_compliance_policy_t policy) {
       return 0;
   }
 }
+
+int SSL_CTX_set_ciphersuites(SSL_CTX *ctx, char *str) {
+  if (!str) {
+    return 0;
+  }
+  bssl::Array<uint8_t> arr;
+  arr.Init(strlen(str));
+  int i = 0;
+  const char *l = str;
+  do {
+    arr[i++] = *l;
+  } while (*++l != '\0');
+  ctx->cipher13_flag_list = std::move(arr);
+  return ctx->cipher13_flag_list.empty() ? 0 : 1;
+}
+
+int SSL_CTX_set_extension_permutation(SSL_CTX *ctx, char *str) {
+  if (!str) {
+    return 0;
+  }
+  bssl::Array<uint8_t> arr;
+  arr.Init(strlen(str));
+  int i = 0;
+  const char *l = str;
+  do {
+    arr[i++] = *l - 1;
+  } while (*++l != '\0');
+  ctx->extension_permutation = std::move(arr);
+  return ctx->extension_permutation.empty() ? 0 : 1;
+}
diff --git a/tool/server.cc b/tool/server.cc
index ebecee03d..a4b29c5aa 100644
--- a/tool/server.cc
+++ b/tool/server.cc
@@ -243,7 +243,7 @@ bool Server(const std::vector<std::string> &args) {
   }
 
   bssl::UniquePtr<SSL_CTX> ctx(SSL_CTX_new(TLS_method()));
-
+  SSL_CTX_set_permute_extensions(ctx.get(), 1);
   const char *keylog_file = getenv("SSLKEYLOGFILE");
   if (keylog_file) {
     g_keylog_file = fopen(keylog_file, "a");
