diff -ruN '--exclude=version.h' '--exclude=buildtest' '--exclude=buildmingw' '--exclude=*.o' pristine/ngtcp2-1.25.0/lib/ngtcp2_fmt.h ngtcp2-1.25.0/lib/ngtcp2_fmt.h
--- pristine/ngtcp2-1.25.0/lib/ngtcp2_fmt.h	2026-08-06 08:23:12.760771800 +0800
+++ ngtcp2-1.25.0/lib/ngtcp2_fmt.h	2026-08-06 08:23:13.201324200 +0800
@@ -74,7 +74,6 @@
     int: ngtcp2_fmt_hex_int_init,                                              \
     short int: ngtcp2_fmt_hex_short_int_init,                                  \
     signed char: ngtcp2_fmt_hex_signed_char_init,                              \
-    char: ngtcp2_fmt_hex_char_init,                                            \
     unsigned long long int: ngtcp2_fmt_hex_init,                               \
     unsigned long int: ngtcp2_fmt_hex_init,                                    \
     unsigned int: ngtcp2_fmt_hex_init,                                         \
@@ -131,7 +130,6 @@
     int: ngtcp2_fmt_hexw_int_init,                                             \
     short int: ngtcp2_fmt_hexw_short_int_init,                                 \
     signed char: ngtcp2_fmt_hexw_signed_char_init,                             \
-    char: ngtcp2_fmt_hexw_char_init,                                           \
     unsigned long long int: ngtcp2_fmt_hexw_init,                              \
     unsigned long int: ngtcp2_fmt_hexw_init,                                   \
     unsigned int: ngtcp2_fmt_hexw_init,                                        \
@@ -218,7 +216,6 @@
     int: ngtcp2_fmt_write_int64,                                               \
     short int: ngtcp2_fmt_write_int64,                                         \
     signed char: ngtcp2_fmt_write_int64,                                       \
-    char: ngtcp2_fmt_write_char,                                               \
     unsigned long long int: ngtcp2_fmt_write_uint64,                           \
     unsigned long int: ngtcp2_fmt_write_uint64,                                \
     unsigned int: ngtcp2_fmt_write_uint64,                                     \
diff -ruN '--exclude=version.h' '--exclude=buildtest' '--exclude=buildmingw' '--exclude=*.o' pristine/ngtcp2-1.25.0/lib/ngtcp2_macro.h ngtcp2-1.25.0/lib/ngtcp2_macro.h
--- pristine/ngtcp2-1.25.0/lib/ngtcp2_macro.h	2026-08-06 08:23:12.792771700 +0800
+++ ngtcp2-1.25.0/lib/ngtcp2_macro.h	2026-08-06 08:23:13.180323400 +0800
@@ -84,7 +84,6 @@
       int: ngtcp2_max_int,                                                     \
       short int: ngtcp2_max_short_int,                                         \
       signed char: ngtcp2_max_signed_char,                                     \
-      char: ngtcp2_max_char,                                                   \
       unsigned long long int: ngtcp2_max_unsigned_long_long_int,               \
       unsigned long int: ngtcp2_max_unsigned_long_int,                         \
       unsigned int: ngtcp2_max_unsigned_int,                                   \
@@ -92,7 +91,6 @@
       unsigned char: ngtcp2_max_unsigned_char),                                \
     short int: ngtcp2_max_short_int,                                           \
     signed char: ngtcp2_max_signed_char,                                       \
-    char: ngtcp2_max_char,                                                     \
     unsigned long long int: ngtcp2_max_unsigned_long_long_int,                 \
     unsigned long int: ngtcp2_max_unsigned_long_int,                           \
     unsigned int: ngtcp2_max_unsigned_int,                                     \
@@ -124,7 +122,6 @@
       int: ngtcp2_min_int,                                                     \
       short int: ngtcp2_min_short_int,                                         \
       signed char: ngtcp2_min_signed_char,                                     \
-      char: ngtcp2_min_char,                                                   \
       unsigned long long int: ngtcp2_min_unsigned_long_long_int,               \
       unsigned long int: ngtcp2_min_unsigned_long_int,                         \
       unsigned int: ngtcp2_min_unsigned_int,                                   \
@@ -132,7 +129,6 @@
       unsigned char: ngtcp2_min_unsigned_char),                                \
     short int: ngtcp2_min_short_int,                                           \
     signed char: ngtcp2_min_signed_char,                                       \
-    char: ngtcp2_min_char,                                                     \
     unsigned long long int: ngtcp2_min_unsigned_long_long_int,                 \
     unsigned long int: ngtcp2_min_unsigned_long_int,                           \
     unsigned int: ngtcp2_min_unsigned_int,                                     \
